### PR TITLE
RSDK-1024-linear-acceleration

### PR DESCRIPTION
**Pull request by susmitaSanyal**
Adding linear acceleration to movementsensor

### CODE REVIEW

File: components/movementsensor/mpu6050/mpu6050.go [hunk 1]
```diff
@@ -283,6 +283,10 @@ func (mpu *mpu6050) LinearVelocity(ctx context.Context, extra map[string]interfa
 	return r3.Vector{}, movementsensor.ErrMethodUnimplementedLinearVelocity
 }
 
+func (mpu *mpu6050) LinearAcceleration(ctx context.Context, exta map[string]interface{}) (r3.Vector, error) {
+	return r3.Vector{}, movementsensor.ErrMethodUnimplementedLinearAcceleration
```
Comment (penguinland) [inline components/movementsensor/mpu6050/mpu6050.go:287]: Like the ADXL345, I kinda want to `return mpu.linearAcceleration, nil` instead. and again, that will require locking the mutex, and checking (and possibly clearing) `mpu.lastError`.

**Review by penguinland — CHANGES_REQUESTED** [review]
My 6 comments are really, like, 3 suggestions repeated twice each.

For sensors that implement the new method and don't just return an Unimplemented error, you should also modify `Properties()` to indicate that they support it. 

When you're ready to try this out on actual hardware, I'd be happy to lend you a breadboard with ADXL and MPU accelerometers wired up (or maybe you already have your own).

**Comment by penguinland** [comment]
I just realized that no one was tagged as a reviewer yet on this! Sorry if I jumped the gun and you weren't ready for a review yet. :confounded:

File: components/movementsensor/imuwit/imu.go [hunk 2]
```diff
@@ -115,8 +115,8 @@ func (imu *wit) Orientation(ctx context.Context, extra map[string]interface{}) (
 	return &imu.orientation, imu.lastError
 }
 
-// GetAcceleration returns accelerometer acceleration in mm_per_sec_per_sec.
-func (imu *wit) GetAcceleration(ctx context.Context) (r3.Vector, error) {
+// LinearAcceleration returns linear acceleration in mm_per_sec_per_sec.
+func (imu *wit) LinearAcceleration(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/imuwit/imu.go:119]: Thanks!

File: components/movementsensor/fake/movementsensor.go [hunk 3]
```diff
@@ -60,6 +60,11 @@ func (f *MovementSensor) LinearVelocity(ctx context.Context, extra map[string]in
 	return r3.Vector{Y: 5.4}, nil
 }
 
+// LinearAcceleration gets the linear acceleration of a fake movementsensor.
+func (f *MovementSensor) LinearAcceleration(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	return r3.Vector{Z: 2}, nil
```
Comment (randhid) [inline components/movementsensor/fake/movementsensor.go:65]: [minor] makes sense to give more than one linear speed field.

**Review by randhid — APPROVED** [review]
Looks good to me. A few minor non-blocking comments.

**Comment by randhid** [comment]
Do you need to change any vue components to see the linear acceleration information in app?

File: components/movementsensor/fake/movementsensor.go (hunk 3 above)
Comment (susmitaSanyal) [reply components/movementsensor/fake/movementsensor.go:65]: Ah yes! good catch.

File: components/movementsensor/client_test.go [hunk 4]
```diff
@@ -152,6 +158,12 @@ func TestClient(t *testing.T) {
 		test.That(t, acc1, test.ShouldResemble, acc)
 		test.That(t, injectMovementSensor.AccuracyFuncExtraCap, test.ShouldResemble, map[string]interface{}{"foo": "bar"})
 
+		la, err := gps1Client.LinearAcceleration(context.Background(), map[string]interface{}{"foo": "bar"})
+		logger.Info(err)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, la, test.ShouldResemble, la)
```
Comment (gvaradarajan) [inline components/movementsensor/client_test.go:164]: it appears this is checking something against itself, did you mean to do this?

File: components/movementsensor/movementsensor_test.go [hunk 5]
```diff
@@ -263,6 +264,7 @@ var (
 	ang     = spatialmath.AngularVelocity{5.5, 1.2, 2.3}
 	orie    = &spatialmath.EulerAngles{5.6, 1.3, 2.4}
 	compass = 123.
+	la      = r3.Vector{0, 0, 0}
```
Comment (gvaradarajan) [inline components/movementsensor/movementsensor_test.go:267]: would it perhaps be better to give this a non-zero value, considering that a zero value is returned in the event of an error or when LinearAcceleration is unimplemented

File: components/movementsensor/adxl345/adxl345.go [hunk 6]
```diff
@@ -251,6 +251,19 @@ func (adxl *adxl345) LinearVelocity(ctx context.Context, extra map[string]interf
 	return r3.Vector{}, movementsensor.ErrMethodUnimplementedLinearVelocity
 }
 
+func (adxl *adxl345) LinearAcceleration(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
```
Comment (gvaradarajan) [inline components/movementsensor/adxl345/adxl345.go:254]: Now that we've implemented `LinearAcceleration` in drivers like these, do these drivers still need to implement readings? Some of them do because they supply information we haven't standardized (like temperature), but otherwise it feels like the readings are now sufficiently handled by the client?

File: components/movementsensor/gpsrtk/gpsrtk.go [hunk 7]
```diff
@@ -567,6 +567,14 @@ func (g *RTKMovementSensor) LinearVelocity(ctx context.Context, extra map[string
 	return g.nmeamovementsensor.LinearVelocity(ctx, extra)
 }
 
+// LinearAcceleration passthrough.
+func (g *RTKMovementSensor) LinearAcceleration(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	if g.lastError != nil {
+		return r3.Vector{}, g.lastError
+	}
+	return g.nmeamovementsensor.LinearAcceleration(ctx, extra)
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtk/gpsrtk.go:575]: just clarifying here, but are we planning in the future to properly implement LinearAcceleration on the nmea sensors? Asking because as of now this appears to defer to a struct that raises the unimplemented error

**Review by gvaradarajan — CHANGES_REQUESTED** [review]
lgtm, just a few questions

File: components/movementsensor/adxl345/adxl345.go (hunk 6 above)
Comment (penguinland) [reply components/movementsensor/adxl345/adxl345.go:254]: My understanding is that every sensor implements `Readings()` to give back all possible readings, and they should continue to do so regardless of what more specialized functions they might also implement.

File: components/movementsensor/gpsrtk/gpsrtk.go (hunk 7 above)
Comment (penguinland) [reply components/movementsensor/gpsrtk/gpsrtk.go:575]: I have no idea if it will ever be implemented, but I like the way this is currently written. If it ever gets implemented, it'll be in the nmea sensor, and then we'll get it here, too, "for free."

File: components/movementsensor/client_test.go [hunk 8]
```diff
@@ -43,7 +43,7 @@ func TestClient(t *testing.T) {
 	heading := 202.
 	props := &movementsensor.Properties{LinearVelocitySupported: true}
 	acc := map[string]float32{"x": 1.1}
-	rs := []interface{}{loc, alt, r3.Vector{0, speed, 0}, r3.Vector{0, 0, ang}, heading, ori}
+	rs := []interface{}{loc, alt, r3.Vector{0, speed, 0}, r3.Vector{0, 0, ang}, r3.Vector{0, 0, 0}, heading, ori}
```
Comment (penguinland) [inline components/movementsensor/client_test.go:46]: I get nervous about adding things into the middle of a list, because anything that was indexing into it will need to be updated (e.g., if there's ever a line like `heading := rs[4]`, it's now broken). Why not add the new part onto the very end?

File: components/movementsensor/client_test.go (hunk 4 above)
Comment (susmitaSanyal) [reply components/movementsensor/client_test.go]: Thanks for catching that!

File: components/movementsensor/movementsensor_test.go [hunk 9]
```diff
@@ -263,6 +264,7 @@ var (
 	ang     = spatialmath.AngularVelocity{5.5, 1.2, 2.3}
 	orie    = &spatialmath.EulerAngles{5.6, 1.3, 2.4}
 	compass = 123.
+	la      = r3.Vector{1.1, 5.6, 0}
```
Comment (penguinland) [inline components/movementsensor/movementsensor_test.go:267]: Reasonable people might disagree, but I think this is too short a name for how broad a scope it has. I got confused in client_test.go about where `la` came from, and had expected it to be local to that function. How about `linAcc`?

**Review by penguinland — APPROVED** [review]
I did not review the Vue or JavaScript changes, but the rest LGTM!

**Comment by susmitaSanyal** [comment]
Will handle vue changes in a separate PR.

File: components/movementsensor/imuwit/imu.go [hunk 10]
```diff
@@ -153,7 +153,7 @@ func (imu *wit) Readings(ctx context.Context, extra map[string]interface{}) (map
 	}
 	readings["magnetometer"] = mag
 
-	acc, err := imu.GetAcceleration(ctx)
+	acc, err := imu.LinearAcceleration(ctx, extra)
 	if err != nil {
 		return nil, err
 	}
```
Comment (penguinland) [inline components/movementsensor/imuwit/imu.go:159]: Oh, shoot. I wish I'd thought to give this feedback earlier. For all sensors that support linear acceleration, we should modify their `Properties()` function (in this file, it's 7 lines below where I'm leaving this comment, because I can't comment that far from one of your changes directly) to indicate that they support linear acceleration. 

For sensors that don't support it, no changes to `Properties()` are needed.

**Review by penguinland — CHANGES_REQUESTED** [review]
Oops! Sorry I forgot about this part before! Now that you've added a new boolean to `movementsensor.Properties`, we need to set it for the sensors that support it.

File: components/movementsensor/cameramono/cameramono.go [hunk 11]
```diff
@@ -290,10 +290,11 @@ func (co *cameramono) AngularVelocity(ctx context.Context, extra map[string]inte
...
 	return &movementsensor.Properties{
-		PositionSupported:        true,
-		OrientationSupported:     true,
-		AngularVelocitySupported: true,
-		LinearVelocitySupported:  true,
+		PositionSupported:           true,
+		OrientationSupported:        true,
+		AngularVelocitySupported:    true,
+		LinearVelocitySupported:     true,
+		LinearAccelerationSupported: true,
```
Comment (penguinland) [inline components/movementsensor/cameramono/cameramono.go:297]: Linear acceleration is not supported on this sensor: it gives back the unimplemented error if you try. Undo this change.

File: components/movementsensor/gpsnmea/pmtkI2C.go [hunk 12]
```diff
@@ -243,8 +243,9 @@ func (g *PmtkI2CNMEAMovementSensor) Orientation(ctx context.Context, extra map[s
 // Properties what can I do!
 func (g *PmtkI2CNMEAMovementSensor) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
 	return &movementsensor.Properties{
-		LinearVelocitySupported: true,
-		PositionSupported:       true,
+		LinearVelocitySupported:     true,
+		PositionSupported:           true,
+		LinearAccelerationSupported: true,
```
Comment (penguinland) [inline components/movementsensor/gpsnmea/pmtkI2C.go:248]: Same thing here: the linear acceleration gives back the "unimplemented" error, so it is not supported here.

**Review by penguinland — COMMENTED** [review]
The ADXL and MPU sensors still need their Properties updated. 

The rule is that if you do something besides returning `movementsensor.ErrMethodUnimplementedLinearVelocity`, that's when you support linear acceleration.

**Review by penguinland — APPROVED** [review]

**Review by gvaradarajan — APPROVED** [review]
